File: apps/web/src/domain/notes/versioning.ts
import { createHash } from "node:crypto";
import type { NoteFile, NoteVersionReason } from "./models";

// Autosaves/checkpoints landing within this window of the latest stored
// autosave/checkpoint row overwrite it in place instead of inserting, so a
// typing burst yields one version instead of one per save.
export const NOTE_VERSION_COALESCE_WINDOW_MS = 60 * 1000;
// Edits whose changed region is at most this many characters are noise
// (a typo fix, one or two letters) and never worth a version on their own.
export const NOTE_VERSION_TRIVIAL_CHAR_DELTA = 2;
// Hard cap on stored versions per note; older rows are pruned on insert so the
// note_versions table (full content snapshots) cannot grow unbounded.
export const NOTE_VERSION_RETENTION_LIMIT = 200;

// Structural on purpose: callers pass a lean Prisma select, not a full
// NoteVersion, so the latest-version check never loads richContent JSON.
export type LatestNoteVersionSnapshot = {
	id: string;
	contentHash: string;
	createdAt: Date;
	content: string;
	reason: string;
} | null;

export type NoteVersionDecision =
	| { action: "skip" }
	| { action: "insert" }
	| { action: "coalesce"; versionId: string };

export type NoteVersionCandidate = Pick<
	NoteFile,
	"name" | "content" | "richContent" | "preferredEditorMode" | "parentId" | "tags" | "properties"
> & {
	reason: NoteVersionReason;
};

export function normalizeVersionTags(tags: string[] | undefined): string[] {
	return [
		...new Set((tags ?? []).map((tag) => tag.trim().toLowerCase()).filter(Boolean)),
	].toSorted((left, right) => left.localeCompare(right));
}

export function buildNoteVersionContentHash(candidate: NoteVersionCandidate): string {
	const payload = JSON.stringify({
		name: candidate.name,
		content: candidate.content,
		richContent: candidate.richContent,
		preferredEditorMode: candidate.preferredEditorMode,
		parentId: candidate.parentId,
		tags: normalizeVersionTags(candidate.tags),
		properties: candidate.properties ?? [],
	});

	return createHash("sha256").update(payload).digest("hex");
}

function countWords(content: string): number {
	return content.split(/\s+/).filter(Boolean).length;
}

function countChars(content: string): number {
	return content.length;
}

// The differing middle of two strings after stripping their common prefix and
// suffix — a cheap stand-in for edit distance that correctly sees same-length
// replacements (where a raw length delta reads as zero change).
function changedRegions(prev: string, next: string): [string, string] {
	let start = 0;
	while (start < prev.length && start < next.length && prev[start] === next[start]) {
		start += 1;
	}
	let prevEnd = prev.length;
	let nextEnd = next.length;
	while (prevEnd > start && nextEnd > start && prev[prevEnd - 1] === next[nextEnd - 1]) {
		prevEnd -= 1;
		nextEnd -= 1;
	}
	return [prev.slice(start, prevEnd), next.slice(start, nextEnd)];
}

function isTrivialContentEdit(prev: string, next: string): boolean {
	if (prev === next) {
		return false;
	}
	const strip = (value: string) => value.replace(/\s+/g, "");
	if (strip(prev) === strip(next)) {
		return true;
	}
	const [prevRegion, nextRegion] = changedRegions(prev, next);
	return Math.max(prevRegion.length, nextRegion.length) <= NOTE_VERSION_TRIVIAL_CHAR_DELTA;
}

function coalescesWith(reason: string): boolean {
	return reason === "autosave" || reason === "checkpoint";
}

/**
 * Decides how to persist a snapshot given the note's latest stored version:
 * identical hash → skip; explicit reasons (created/rename/restore) → insert;
 * whitespace-only or ≤2-char edits → skip; within the coalesce window of the
 * latest autosave/checkpoint row → overwrite that row; otherwise insert.
 */
export function decideNoteVersionPersistence(
	candidate: NoteVersionCandidate & { createdAt: Date },
	latestVersion: LatestNoteVersionSnapshot,
): NoteVersionDecision {
	if (!latestVersion) {
		return { action: "insert" };
	}

	const nextHash = buildNoteVersionContentHash(candidate);
	if (nextHash === latestVersion.contentHash) {
		return { action: "skip" };
	}

	if (!coalescesWith(candidate.reason)) {
		return { action: "insert" };
	}

	if (isTrivialContentEdit(latestVersion.content, candidate.content)) {
		return { action: "skip" };
	}

	const elapsed = candidate.createdAt.getTime() - latestVersion.createdAt.getTime();
	if (coalescesWith(latestVersion.reason) && elapsed < NOTE_VERSION_COALESCE_WINDOW_MS) {
		return { action: "coalesce", versionId: latestVersion.id };
	}

	return { action: "insert" };
}

export function summarizeNoteVersionReason(reason: NoteVersionReason): string {
	switch (reason) {
		case "created":
			return "Initial version";
		case "rename":
			return "Metadata update";
		case "restore":
			return "Restored checkpoint";
		case "checkpoint":
			return "Checkpoint";
		case "autosave":
		default:
			return "Autosaved checkpoint";
	}
}

export type NoteVersionDeltaValues = {
	wordDelta: number;
	charDelta: number;
};

export function getNoteVersionDeltaValues(
	currentContent: string,
	previousContent?: string,
): NoteVersionDeltaValues | null {
	if (!previousContent) {
		return null;
	}

	const currentWords = countWords(currentContent);
	const previousWords = countWords(previousContent);
	const currentChars = countChars(currentContent);
	const previousChars = countChars(previousContent);

	return {
		wordDelta: currentWords - previousWords,
		charDelta: currentChars - previousChars,
	};
}

export function formatNoteVersionDelta(currentContent: string, previousContent?: string): string {
	const values = getNoteVersionDeltaValues(currentContent, previousContent);
	if (!values) {
		return "+0 -0";
	}

	const wordLabel = values.wordDelta >= 0 ? `+${values.wordDelta}` : `${values.wordDelta}`;
	const charLabel = values.charDelta >= 0 ? `+${values.charDelta}` : `${values.charDelta}`;

	return `${wordLabel} ${charLabel}`;
}

export function previewVersionContent(content: string): string[] {
	return content
		.split("\n")
		.map((line) => line.trimEnd())
		.filter((line) => line.trim().length > 0)
		.slice(0, 3);
}

export function getVersionContextPreview(content: string, maxLength = 52): string | null {
	const lines = previewVersionContent(content);
	if (lines.length === 0) return null;

	const first = lines[0].replace(/^#+\s+/, "").trim();
	if (!first) return null;

	if (first.length <= maxLength) return first;
	return `${first.slice(0, maxLength - 1).trimEnd()}…`;
}

File: v1/apps/desktop/src-tauri/src/versioning.rs
//! Note version capture/dedupe rules. Mirrors
//! `apps/web/src/domain/notes/versioning.ts` so a desktop checkpoint behaves the
//! same as a web one — same trivial-edit filter, same coalesce window, same
//! retention cap. The content hash here only needs to be internally
//! consistent (it dedupes within this SQLite table), not byte-identical to the
//! web hash, since desktop versions are local-only and never synced.

use sha2::{Digest, Sha256};

use crate::storage::NoteVersionSnapshot;

/// Autosaves/checkpoints landing within this window of the latest stored
/// autosave/checkpoint row overwrite it in place instead of inserting, so a
/// typing burst yields one version instead of one per save.
pub const COALESCE_WINDOW_MS: i64 = 60 * 1000;
/// Edits whose changed region is at most this many characters are noise
/// (a typo fix, one or two letters) and never worth a version on their own.
pub const TRIVIAL_CHAR_DELTA: usize = 2;
/// Hard cap on stored versions per note; older rows are pruned on insert so the
/// note_versions table (full content snapshots) cannot grow unbounded.
pub const RETENTION_LIMIT: i64 = 200;

/// The lean fields needed to decide whether a new snapshot is worth persisting,
/// without loading the latest version's full `richContent`/properties JSON.
pub struct LatestVersionLean {
    pub id: String,
    pub content_hash: String,
    pub created_at: i64,
    pub content: String,
    pub reason: String,
}

/// What to do with an incoming snapshot relative to the latest stored version.
#[derive(Debug, PartialEq, Eq)]
pub enum VersionDecision {
    Skip,
    Insert,
    /// Overwrite the latest row (by id) in place.
    Coalesce(String),
}

fn normalize_tags(tags: &[String]) -> Vec<String> {
    let mut normalized: Vec<String> = tags
        .iter()
        .map(|tag| tag.trim().to_lowercase())
        .filter(|tag| !tag.is_empty())
        .collect();
    normalized.sort();
    normalized.dedup();
    normalized
}

pub fn content_hash(snapshot: &NoteVersionSnapshot) -> String {
    let payload = serde_json::json!({
        "name": snapshot.name,
        "content": snapshot.content,
        "richContent": snapshot.rich_content,
        "preferredEditorMode": snapshot.preferred_editor_mode,
        "parentId": snapshot.parent_id,
        "tags": normalize_tags(&snapshot.tags),
        "properties": snapshot.properties,
    });
    let serialized = serde_json::to_string(&payload).unwrap_or_default();
    let mut hasher = Sha256::new();
    hasher.update(serialized.as_bytes());
    format!("{:x}", hasher.finalize())
}

/// The differing middle of two strings after stripping their common prefix and
/// suffix — a cheap stand-in for edit distance that correctly sees same-length
/// replacements (where a raw length delta reads as zero change).
fn changed_regions(prev: &str, next: &str) -> (Vec<char>, Vec<char>) {
    let prev: Vec<char> = prev.chars().collect();
    let next: Vec<char> = next.chars().collect();
    let mut start = 0;
    while start < prev.len() && start < next.len() && prev[start] == next[start] {
        start += 1;
    }
    let mut prev_end = prev.len();
    let mut next_end = next.len();
    while prev_end > start && next_end > start && prev[prev_end - 1] == next[next_end - 1] {
        prev_end -= 1;
        next_end -= 1;
    }
    (
        prev[start..prev_end].to_vec(),
        next[start..next_end].to_vec(),
    )
}

fn is_trivial_content_edit(prev: &str, next: &str) -> bool {
    if prev == next {
        return false;
    }
    let strip = |s: &str| s.chars().filter(|c| !c.is_whitespace()).collect::<String>();
    if strip(prev) == strip(next) {
        return true;
    }
    let (prev_region, next_region) = changed_regions(prev, next);
    prev_region.len().max(next_region.len()) <= TRIVIAL_CHAR_DELTA
}

fn coalesces_with(reason: &str) -> bool {
    matches!(reason, "autosave" | "checkpoint")
}

/// Decides how to persist `snapshot` given the most recently stored version
/// (`latest`, or `None` for a note with no history yet):
/// - identical content hash → skip
/// - explicit reasons (created/rename/restore) → always insert on change
/// - whitespace-only or ≤`TRIVIAL_CHAR_DELTA`-char edits → skip
/// - within `COALESCE_WINDOW_MS` of the latest autosave/checkpoint row →
///   overwrite that row instead of inserting
/// - otherwise → insert
pub fn decide(
    snapshot: &NoteVersionSnapshot,
    reason: &str,
    created_at: i64,
    latest: Option<&LatestVersionLean>,
) -> VersionDecision {
    let Some(latest) = latest else {
        return VersionDecision::Insert;
    };

    if content_hash(snapshot) == latest.content_hash {
        return VersionDecision::Skip;
    }

    if !coalesces_with(reason) {
        return VersionDecision::Insert;
    }

    if is_trivial_content_edit(&latest.content, &snapshot.content) {
        return VersionDecision::Skip;
    }

    let elapsed = created_at - latest.created_at;
    if coalesces_with(&latest.reason) && elapsed < COALESCE_WINDOW_MS {
        return VersionDecision::Coalesce(latest.id.clone());
    }

    VersionDecision::Insert
}

#[cfg(test)]
mod tests {
    use super::*;

    fn snapshot(content: &str) -> NoteVersionSnapshot {
        NoteVersionSnapshot {
            name: "Note.md".to_string(),
            content: content.to_string(),
            rich_content: serde_json::json!([]),
            preferred_editor_mode: "block".to_string(),
            parent_id: None,
            tags: vec![],
            properties: serde_json::json!([]),
        }
    }

    fn latest(content: &str, reason: &str, created_at: i64) -> LatestVersionLean {
        LatestVersionLean {
            id: "v1".to_string(),
            content_hash: content_hash(&snapshot(content)),
            created_at,
            content: content.to_string(),
            reason: reason.to_string(),
        }
    }

    #[test]
    fn no_latest_always_inserts() {
        assert_eq!(
            decide(&snapshot("hello"), "autosave", 0, None),
            VersionDecision::Insert
        );
    }

    #[test]
    fn identical_hash_never_persists_regardless_of_reason() {
        let l = latest("hello", "autosave", 0);
        assert_eq!(
            decide(&snapshot("hello"), "rename", 1, Some(&l)),
            VersionDecision::Skip
        );
        assert_eq!(
            decide(&snapshot("hello"), "checkpoint", 1, Some(&l)),
            VersionDecision::Skip
        );
    }

    #[test]
    fn non_autosave_reason_always_inserts_on_change() {
        let l = latest("hello", "autosave", 0);
        assert_eq!(
            decide(&snapshot("hello world"), "rename", 1, Some(&l)),
            VersionDecision::Insert
        );
    }

    #[test]
    fn whitespace_only_edit_is_skipped() {
        let l = latest("hello world", "autosave", 0);
        assert_eq!(
            decide(
                &snapshot("hello  world\n"),
                "autosave",
                COALESCE_WINDOW_MS * 10,
                Some(&l)
            ),
            VersionDecision::Skip
        );
    }

    #[test]
    fn one_or_two_char_edit_is_skipped() {
        let l = latest("hello", "autosave", 0);
        assert_eq!(
            decide(
                &snapshot("hello!"),
                "autosave",
                COALESCE_WINDOW_MS * 10,
                Some(&l)
            ),
            VersionDecision::Skip
        );
        assert_eq!(
            decide(
                &snapshot("heLLo"),
                "autosave",
                COALESCE_WINDOW_MS * 10,
                Some(&l)
            ),
            VersionDecision::Skip
        );
    }

    #[test]
    fn same_length_replacement_is_seen_as_change() {
        let l = latest("the cat sat", "autosave", 0);
        assert_eq!(
            decide(
                &snapshot("the dog ran"),
                "autosave",
                COALESCE_WINDOW_MS,
                Some(&l)
            ),
            VersionDecision::Insert
        );
    }

    #[test]
    fn meaningful_edit_within_window_coalesces() {
        let l = latest("hello", "autosave", 0);
        assert_eq!(
            decide(
                &snapshot("hello brave new world"),
                "autosave",
                1_000,
                Some(&l)
            ),
            VersionDecision::Coalesce("v1".to_string())
        );
    }

    #[test]
    fn meaningful_edit_after_window_inserts() {
        let l = latest("hello", "autosave", 0);
        assert_eq!(
            decide(
                &snapshot("hello brave new world"),
                "autosave",
                COALESCE_WINDOW_MS,
                Some(&l)
            ),
            VersionDecision::Insert
        );
    }

    #[test]
    fn never_coalesces_into_explicit_versions() {
        let l = latest("hello", "restore", 0);
        assert_eq!(
            decide(
                &snapshot("hello brave new world"),
                "autosave",
                1_000,
                Some(&l)
            ),
            VersionDecision::Insert
        );
    }
}
